### PR TITLE
Use SecretString in AuthInfo to avoid credential leaking

### PR DESCRIPTION
**Pull request by ChinYing-Li**
## Motivation
This PR aims to solve #751.

## Solution
the `secrecy` crate is used, and `AuthInfo.password` is now of type `Option<SecretString>`. Are there other structs/fields that would require secrets?
Since `SecretString` has no `PartialEq` implementation, there's a custom `impl PartialEq for AuthInfo`, which is a hack that compares serialized structs. Please let me know if such a hack is not acceptable.
Any suggestion is appreciated!

### CODE REVIEW

File: kube-client/src/config/file_config.rs [hunk 1]
```diff
@@ -118,16 +119,40 @@ pub struct NamedAuthInfo {
...
-    pub password: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none", default)]
+    #[serde(
+        serialize_with = "serialize_password",
+        deserialize_with = "deserialize_password"
+    )]
+    pub password: Option<SecretString>,
 
     /// The bearer token for authentication to the kubernetes cluster.
     #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:158]: `token` should be `Option<SecretString>` as well.

File: kube-client/src/config/file_config.rs [hunk 2]
```diff
@@ -627,4 +659,26 @@ users:
...
         assert_eq!(cfg, Kubeconfig::default());
     }
+
+    #[test]
+    fn authinfo_debug_does_not_output_password() {
+        let authinfo_yaml = r#"
+username: user
+password: kube_rs
+"#;
+        let authinfo: AuthInfo = serde_yaml::from_str(&authinfo_yaml).unwrap();
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:669]:
```suggestion
        let authinfo: AuthInfo = serde_yaml::from_str(authinfo_yaml).unwrap();
```

File: kube-client/src/config/file_config.rs [hunk 3]
```diff
@@ -118,16 +119,40 @@ pub struct NamedAuthInfo {
...
+        None => serializer.serialize_none(),
+    }
+}
+
+fn deserialize_password<'de, D>(deserializer: D) -> Result<Option<SecretString>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let buffer = String::deserialize(deserializer);
+    match buffer {
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:137]:
```suggestion
    match String::deserialize(deserializer) {
```

File: kube-client/src/config/file_config.rs [hunk 4]
```diff
@@ -174,6 +199,13 @@ pub struct AuthInfo {
     pub exec: Option<ExecConfig>,
 }
 
+#[cfg(test)]
+impl PartialEq for AuthInfo {
+    fn eq(&self, other: &Self) -> bool {
+        serde_json::to_value(&self).unwrap() == serde_json::to_value(&other).unwrap()
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:205]: I think this is fine. It was added recently for `kubeconfig_from_empty_string` test.

**Review by kazk — COMMENTED** [review]
Thanks for doing this!

I think `token` field should be `Option<SecretString>` as well.

There's also `client_key_data`, but we might want to clean up around `identity_pem` first in a separate PR.

**Comment by ChinYing-Li** [comment]
@kazk Seems like no one else is working on refactoring `identity_pem`, perhaps I can give it a try in a separate PR?

**Comment by kazk** [comment]
@ChinYing-Li Thanks, but I've been looking into it already (originally started looking into it because I have an idea for #542 and working with concatenated `identity_pem` is not ideal), and planning to work on it after #768.

As far as I can tell, the concatenated `identity_pem` shouldn't be necessary. At least, it shouldn't be necessary to be stored in `Config` because the client certificates and the private key are in `config.auth_info`:

https://github.com/kube-rs/kube-rs/blob/52f69b9a89aa9e23605eb7f9c3ca673260e21853/kube-client/src/config/file_loader.rs#L99-L105

But we'll need to look into _why_ we're doing the following:

https://github.com/kube-rs/kube-rs/blob/52f69b9a89aa9e23605eb7f9c3ca673260e21853/kube-client/src/config/mod.rs#L269-L278

to set `accept_invalid_certs` only when loading `identity_pem` fails.

If there's no problem with doing
```rust
accept_invalid_certs = loader.cluster.insecure_skip_tls_verify.unwrap_or(false);
```
then, I think we can just get rid of `identity_pem` and rewrite TLS methods to accept the cert chain and the private key separately.

**Comment by kazk** [comment]
Opened #771 that removes `identity_pem` with minimal changes

**Comment by kazk** [comment]
@ChinYing-Li #771 is merged, so you can do something like the following after rebasing:

```diff
diff --git i/kube-client/src/config/file_config.rs w/kube-client/src/config/file_config.rs
index a824699..4e16f90 100644
--- i/kube-client/src/config/file_config.rs
+++ w/kube-client/src/config/file_config.rs
@@ -181,7 +181,11 @@ pub struct AuthInfo {
     /// PEM-encoded data from a client key file for TLS. Overrides `client_key`
     #[serde(rename = "client-key-data")]
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub client_key_data: Option<String>,
+    #[serde(
+        serialize_with = "serialize_secretstring",
+        deserialize_with = "deserialize_secretstring"
+    )]
+    pub client_key_data: Option<SecretString>,
 
     /// The username to act-as.
     #[serde(rename = "as")]
@@ -430,8 +434,11 @@ impl Cluster {
             return Ok(None);
         }
 
-        let ca = load_from_base64_or_file(&self.certificate_authority_data, &self.certificate_authority)
-            .map_err(KubeconfigError::LoadCertificateAuthority)?;
+        let ca = load_from_base64_or_file(
+            &self.certificate_authority_data.as_deref(),
+            &self.certificate_authority,
+        )
+        .map_err(KubeconfigError::LoadCertificateAuthority)?;
         Ok(Some(ca))
     }
 }
@@ -448,24 +455,26 @@ impl AuthInfo {
     pub(crate) fn load_client_certificate(&self) -> Result<Vec<u8>, KubeconfigError> {
         // TODO Shouldn't error when `self.client_certificate_data.is_none() && self.client_certificate.is_none()`
 
-        load_from_base64_or_file(&self.client_certificate_data, &self.client_certificate)
+        load_from_base64_or_file(&self.client_certificate_data.as_deref(), &self.client_certificate)
             .map_err(KubeconfigError::LoadClientCertificate)
     }
 
     pub(crate) fn load_client_key(&self) -> Result<Vec<u8>, KubeconfigError> {
         // TODO Shouldn't error when `self.client_key_data.is_none() && self.client_key.is_none()`
 
-        load_from_base64_or_file(&self.client_key_data, &self.client_key)
-            .map_err(KubeconfigError::LoadClientKey)
+        load_from_base64_or_file(
+            &self.client_key_data.as_ref().map(|s| s.expose_secret().as_str()),
+            &self.client_key,
+        )
+        .map_err(KubeconfigError::LoadClientKey)
     }
 }
 
 fn load_from_base64_or_file<P: AsRef<Path>>(
-    value: &Option<String>,
+    value: &Option<&str>,
     file: &Option<P>,
 ) -> Result<Vec<u8>, LoadDataError> {
     let data = value
-        .as_deref()
         .map(load_from_base64)
         .or_else(|| file.as_ref().map(load_from_file))
         .unwrap_or(Err(LoadDataError::NoBase64DataOrFile))?;
```

**Comment by ChinYing-Li** [comment]
I have updated the PR, so now `client_key_data` is also `Option<SecretString>. PR ready for review!

File: kube-client/Cargo.toml [hunk 5]
```diff
@@ -69,6 +69,7 @@ hyper-timeout = {version = "0.4.1", optional = true }
 tame-oauth = { version = "0.6.0", features = ["gcp"], optional = true }
 pin-project = { version = "1.0.4", optional = true }
 rand = { version = "0.8.3", optional = true }
+secrecy = { version = "0.8.0", features = ['alloc', 'serde'] }
```
Comment (kazk) [inline kube-client/Cargo.toml:72]: nit: consistency
```suggestion
secrecy = { version = "0.8.0", features = ["alloc", "serde"] }
```

File: kube-client/src/config/file_config.rs [hunk 6]
```diff
@@ -118,20 +119,47 @@ pub struct NamedAuthInfo {
...
+
+fn deserialize_secretstring<'de, D>(deserializer: D) -> Result<Option<SecretString>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    match String::deserialize(deserializer) {
+        Ok(pw_string) => Ok(Some(SecretString::new(pw_string))),
+        Err(e) => Err(e),
+    }
+}
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:140]: nit: I'd change `pw_string` to `secret`.

**Review by kazk — APPROVED** [review]
Looks good to me. Just a few minor nits.

**Comment by kazk** [comment]
Going to merge this to unblock #768. Thanks again.

**Pull request merged**